From: Tommy Webb <tommy@calyxinstitute.org>
Date: Mon, 29 Jul 2024 20:18:04 +0000
Subject: UX changes to "Add custom tab intents privacy option"

Remove the "Allow custom tab intents" toggle. It is now tied to
the "Open external links in incognito" toggle.

This is done for a few reasons:
1. Turning off Custom Tabs seems fairly niche, and we only initially
   wanted to add the external-links-in-incognito part.
2. We accidentally kept Cromite's default-disabled status for the
   Custom Tabs toggle and don't have a great way to migrate out.
3. Custom Tabs are supposed to be turned off when external incognito
   links is on. This is the behavior described but not what happens.
   Activity from third-party apps via Custom Tabs could unexpectedly
   leak into the user's browser history. Simply removing the separate
   toggle fixes this automatically, no UI changes needed.

Change-Id: Id11bab82a3a634080165bb51a3533f782bd6ccad
---
 .../android/java/res/xml/privacy_preferences.xml  |  9 ++-------
 .../chrome/browser/LaunchIntentDispatcher.java    |  5 +++--
 .../browser/privacy/settings/PrivacySettings.java | 11 -----------
 .../ui/android/strings/android_chrome_strings.grd | 15 ++++-----------
 4 files changed, 9 insertions(+), 31 deletions(-)

diff --git a/chrome/android/java/res/xml/privacy_preferences.xml b/chrome/android/java/res/xml/privacy_preferences.xml
--- a/chrome/android/java/res/xml/privacy_preferences.xml
+++ b/chrome/android/java/res/xml/privacy_preferences.xml
@@ -60,15 +60,10 @@ found in the LICENSE file.
         android:key="gpc"
         android:title="@string/gpc_title"
         android:fragment="org.chromium.chrome.browser.privacy.settings.GpcSettings" />
-    <org.chromium.components.browser_ui.settings.ChromeSwitchPreference
-        android:key="allow_custom_tab_intents"
-        android:title="@string/allow_custom_tab_intents_title"
-        android:summary="@string/allow_custom_tab_intents_summary"
-        android:defaultValue="false" />
     <org.chromium.components.browser_ui.settings.ChromeSwitchPreference
         android:key="open_external_links_incognito"
-        android:title="@string/open_external_links_incognito_title"
-        android:summary="@string/open_external_links_incognito_summary"
+        android:title="@string/open_external_links_incognito_combined_title"
+        android:summary="@string/open_external_links_incognito_combined_summary"
         android:defaultValue="false" />
     <org.chromium.chrome.browser.about_settings.HyperlinkPreference
         android:key="cromite_flags"
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/LaunchIntentDispatcher.java b/chrome/android/java/src/org/chromium/chrome/browser/LaunchIntentDispatcher.java
--- a/chrome/android/java/src/org/chromium/chrome/browser/LaunchIntentDispatcher.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/LaunchIntentDispatcher.java
@@ -242,9 +242,10 @@ public class LaunchIntentDispatcher {
      */
     public static boolean isCustomTabIntent(Intent intent) {
         if (intent == null) return false;
-        if (!ContextUtils.getAppSharedPreferences()
-                .getBoolean(PrivacySettings.PREF_ALLOW_CUSTOM_TAB_INTENTS, false))
+        if (ContextUtils.getAppSharedPreferences()
+                .getBoolean(PrivacySettings.PREF_OPEN_EXTERNAL_LINKS_INCOGNITO, false)) {
             return false;
+        }
         Log.w(
                 TAG,
                 "CustomTabsIntent#shouldAlwaysUseBrowserUI() = "
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
--- a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
@@ -98,7 +98,6 @@ public class PrivacySettings extends ChromeBaseSettingsFragment
     private IncognitoLockSettings mIncognitoLockSettings;
     private SettingsLauncher mSettingsLauncher;
 
-    private ChromeSwitchPreference allowCustomTabIntentsPref;
     private ChromeSwitchPreference openExternalLinksPref;
 
     @Override
@@ -319,7 +318,6 @@ public class PrivacySettings extends ChromeBaseSettingsFragment
                 new SpanApplier.SpanInfo("<link2>", "</link2>", servicesLink));
     }
 
-    public static final String PREF_ALLOW_CUSTOM_TAB_INTENTS = "allow_custom_tab_intents";
     public static final String PREF_OPEN_EXTERNAL_LINKS_INCOGNITO = "open_external_links_incognito";
 
     @Override
@@ -334,10 +332,6 @@ public class PrivacySettings extends ChromeBaseSettingsFragment
         } else if (PREF_SEARCH_SUGGESTIONS.equals(key)) {
             UserPrefs.get(getProfile())
                     .setBoolean(Pref.SEARCH_SUGGEST_ENABLED, (boolean) newValue);
-        } else if (PREF_ALLOW_CUSTOM_TAB_INTENTS.equals(key)) {
-            SharedPreferences.Editor sharedPreferencesEditor = ContextUtils.getAppSharedPreferences().edit();
-            sharedPreferencesEditor.putBoolean(PREF_ALLOW_CUSTOM_TAB_INTENTS, (boolean)newValue);
-            sharedPreferencesEditor.apply();
         } else if (PREF_OPEN_EXTERNAL_LINKS_INCOGNITO.equals(key)) {
             SharedPreferences.Editor sharedPreferencesEditor = ContextUtils.getAppSharedPreferences().edit();
             sharedPreferencesEditor.putBoolean(PREF_OPEN_EXTERNAL_LINKS_INCOGNITO, (boolean)newValue);
@@ -364,11 +358,6 @@ public class PrivacySettings extends ChromeBaseSettingsFragment
                     UserPrefs.get(getProfile()).getBoolean(Pref.CAN_MAKE_PAYMENT_ENABLED));
         }
 
-        allowCustomTabIntentsPref =
-                (ChromeSwitchPreference) findPreference(PREF_ALLOW_CUSTOM_TAB_INTENTS);
-        allowCustomTabIntentsPref.setOnPreferenceChangeListener(this);
-        allowCustomTabIntentsPref.setManagedPreferenceDelegate(mManagedPreferenceDelegate);
-
         openExternalLinksPref =
                 (ChromeSwitchPreference) findPreference(PREF_OPEN_EXTERNAL_LINKS_INCOGNITO);
         openExternalLinksPref.setOnPreferenceChangeListener(this);
diff --git a/chrome/browser/ui/android/strings/android_chrome_strings.grd b/chrome/browser/ui/android/strings/android_chrome_strings.grd
--- a/chrome/browser/ui/android/strings/android_chrome_strings.grd
+++ b/chrome/browser/ui/android/strings/android_chrome_strings.grd
@@ -5475,19 +5475,12 @@ To change this setting, <ph name="BEGIN_LINK">BEGIN_LINK</ph>delete the Chrome d
       <message name="IDS_NEAR_OOM_INTERVENTION_DECLINE" desc="The text of the button letting the user decline the browser's intervention, so that the page can resume what it was doing.">
          Resume
       </message>
-      <!-- Allow custom tab intents -->
-      <message name="IDS_ALLOW_CUSTOM_TAB_INTENTS_TITLE" desc="Text for 'Allow custom tab intents' settings-privacy option.">
-        Allow custom tab intents
-      </message>
-      <message name="IDS_ALLOW_CUSTOM_TAB_INTENTS_SUMMARY" desc="Summary text for 'Allow custom tab intents' settings-privacy option.">
-        Allow applications to open custom tab intents, similar to webview.
-      </message>
-      <!-- Open External Links in Incognito -->
-      <message name="IDS_OPEN_EXTERNAL_LINKS_INCOGNITO_TITLE" desc="Text for 'Open external links in incognito' settings-privacy option.">
+      <!-- Open External Links in Incognito and turn off Custom Tabs -->
+      <message name="IDS_OPEN_EXTERNAL_LINKS_INCOGNITO_COMBINED_TITLE" desc="Text for 'Open external links in incognito' settings-privacy option.">
         Open external links in incognito
       </message>
-      <message name="IDS_OPEN_EXTERNAL_LINKS_INCOGNITO_SUMMARY" desc="Summary text for 'Open external links in incognito' settings-privacy option.">
-        Force the opening of all external links in incognito mode
+      <message name="IDS_OPEN_EXTERNAL_LINKS_INCOGNITO_COMBINED_SUMMARY" desc="Summary text for 'Open external links in incognito' settings-privacy option.">
+        Force the opening of all external links in incognito mode. Also turns off Custom Tabs
       </message>
       <!-- Usage Stats strings -->
       <message name="IDS_USAGE_STATS_CONSENT_TITLE" desc="Title for activity authorizing Digital Wellbeing to access Chrome usage data">
-- 

